### PR TITLE
chore: add knip

### LINKED ISSUE #8021: Repo: Add Knip
### Suggestion

As mentioned in https://github.com/typescript-eslint/typescript-eslint/issues/371#issuecomment-1427383474 & https://github.com/typescript-eslint/typescript-eslint/issues/371#issuecomment-1770092707: https://github.com/webpro/knip is fantastic. It does unused exports detection as well as dependency analysis and a few other checks. Let's add it to typescript-eslint to find dead code!

### PR DESCRIPTION
**Pull request by auvred**
## PR Checklist

- [x] Addresses an existing open issue: fixes #8021
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

TODO:
- [x] - Fix knip errors in all packages:
  - [x] . (root)
    - [x] - https://github.com/webpro/knip/issues/433
    - [x] - https://github.com/webpro/knip/pull/432
  - [x] ast-spec
  - [x] eslint-plugin
  - [x] eslint-plugin-internal
  - [x] eslint-plugin-tslint
  - [x] integration-tests
  - [x] parser
  - [x] repo-tools
  - [x] rule-schema-to-typescript-types
  - [x] rule-tester
  - [x] scope-manager
  - [x] type-utils
  - [x] types
  - [x] typescript-estree
  - [x] utils
  - [x] visitor-keys
  - [x] website
    - [x] TODO: sync with https://github.com/typescript-eslint/typescript-eslint/pull/8209
  - [x] website-eslint
- [x] - Write a short review explaining non-obvious changes

### CODE REVIEW

**Comment by JoshuaKGoldberg** [comment]
Now that the flat config work is in, I think we're about ready to seriously look at this 🚀. For the remaining bits not supported by Knip yet, I'd say let's add disables in the `knip.json` with comments pointing to tracking issues. I'd rather get the tool in sooner than wait for perfection.

File: package.json
Comment (JoshuaKGoldberg) [inline package.json:104]: > Knip has a lot of dependencies (200+) ([npmgraph](https://npmgraph.js.org/?q=knip)). Maybe we should only install it in CI?

Personally, I don't mind. The lockfile is pretty huge and uses nx/Yarn caching anyway. If we want to make the story for large devDependencies better we'd probably want to investigate a more holistic strategy. But I won't hold this opinion strongly if anybody has a strong preference.

FWIW a `yarn dedupe` does remove a few entries... not too many though.

File: packages/repo-tools/package.json
Comment (armano2) [inline packages/repo-tools/package.json:23]:
```suggestion
    "@typescript-eslint/scope-manager": "6.19.0",
    "@typescript-eslint/types": "6.19.0",
    "@typescript-eslint/typescript-estree": "6.19.0",
    "@typescript-eslint/utils": "6.19.0",
```

File: package.json
Comment (auvred) [reply package.json:104]: Makes sense! Thanks!

File: packages/eslint-plugin/package.json
Comment (auvred) [inline packages/eslint-plugin/package.json:69]: Unused after https://github.com/typescript-eslint/typescript-eslint/pull/3658

File: packages/eslint-plugin/package.json
Comment (auvred) [inline packages/eslint-plugin/package.json:73]: Unused after https://github.com/typescript-eslint/typescript-eslint/pull/8377

File: packages/eslint-plugin/package.json
Comment (auvred) [inline packages/eslint-plugin/package.json:84]: Unused after https://github.com/typescript-eslint/typescript-eslint/commit/d6766838a#diff-26a6efddd9bcd7d58918df668665bc51f5e57d5bf30cbe90f9b8c33fc792154d

File: packages/eslint-plugin/package.json
Comment (auvred) [inline packages/eslint-plugin/package.json:88]: Unused after https://github.com/typescript-eslint/typescript-eslint/commit/faea3ff8b

File: packages/eslint-plugin/package.json
Comment (auvred) [inline packages/eslint-plugin/package.json:85]: Required by https://github.com/typescript-eslint/typescript-eslint/blob/e44a1a280f08f9fd0d29f74e5c3e73b7b64a9606/packages/eslint-plugin/tests/rules/no-use-before-define.test.ts#L721

File: knip.ts
Comment (auvred) [inline knip.ts:65]: Knip thinks that `tsconfig.json` is unlisted dependency.

We could change 
```diff
-    "typecheck": "npx tsc -p tsconfig.json --noEmit"
+    "typecheck": "tsc -p tsconfig.json --noEmit"
```

in 

https://github.com/typescript-eslint/typescript-eslint/blob/e44a1a280f08f9fd0d29f74e5c3e73b7b64a9606/packages/repo-tools/package.json#L16

to avoid this error.

But I'm not sure if we can safely do this because `npx` was intentionally added in https://github.com/typescript-eslint/typescript-eslint/pull/8450

File: packages/rule-tester/package.json
Comment (auvred) [inline packages/rule-tester/package.json:66]: https://github.com/typescript-eslint/typescript-eslint/blob/e44a1a280f08f9fd0d29f74e5c3e73b7b64a9606/packages/rule-tester/tests/eslint-base/fixtures/enhanced-parser2.js#L6

File: packages/rule-tester/package.json
Comment (auvred) [inline packages/rule-tester/package.json:67]: https://github.com/typescript-eslint/typescript-eslint/blob/e44a1a280f08f9fd0d29f74e5c3e73b7b64a9606/packages/rule-tester/tests/eslint-base/eslint-base.test.js#L27

File: packages/rule-tester/package.json
Comment (auvred) [inline packages/rule-tester/package.json:68]: https://github.com/typescript-eslint/typescript-eslint/blob/e44a1a280f08f9fd0d29f74e5c3e73b7b64a9606/packages/rule-tester/tests/eslint-base/eslint-base.test.js#L1050

File: packages/typescript-estree/tests/test-utils/tserror-serializer.ts
Comment (auvred) [inline packages/typescript-estree/tests/test-utils/tserror-serializer.ts:1]: This file is unused after https://github.com/typescript-eslint/typescript-eslint/pull/6612

File: packages/utils/src/eslint-utils/index.ts
Comment (auvred) [inline packages/utils/src/eslint-utils/index.ts:2]: This import was removed in https://github.com/typescript-eslint/typescript-eslint/pull/8377, but wasn't restored in https://github.com/typescript-eslint/typescript-eslint/pull/8399

This is a publicly visible change. Maybe we should fix it in separate PR?

File: packages/utils/package.json
Comment (auvred) [inline packages/utils/package.json:71]: It was forked in https://github.com/typescript-eslint/typescript-eslint/pull/6963, so now it's unused

File: packages/utils/package.json
Comment (auvred) [inline packages/utils/package.json:72]: This line was initially removed in https://github.com/typescript-eslint/typescript-eslint/commit/2ce1c1d22#diff-21f8eb1d7a149d42ab0454ea905d44f624ff1d0aeb6d7a88d75b42962ca6b053

File: packages/utils/package.json
Comment (auvred) [inline packages/utils/package.json:76]: Unused after https://github.com/typescript-eslint/typescript-eslint/commit/c33f497ad#diff-e1edab5ff628076cdee7df126c59ba2f7b2e5d1e5a4a5671343ab437a4f9ffdb

File: packages/website/src/components/hooks/useFocus.ts
Comment (auvred) [inline packages/website/src/components/hooks/useFocus.ts:1]: Unused after https://github.com/typescript-eslint/typescript-eslint/commit/efd6ca9f1

File: packages/website/src/hooks/useMediaQuery.ts
Comment (auvred) [inline packages/website/src/hooks/useMediaQuery.ts:1]: Unused after https://github.com/typescript-eslint/typescript-eslint/commit/0d8d9f295

File: packages/website-eslint/package.json
Comment (auvred) [inline packages/website-eslint/package.json:42]: Unused after https://github.com/typescript-eslint/typescript-eslint/pull/6716

File: packages/website-eslint/package.json
Comment (auvred) [inline packages/website-eslint/package.json:28]: Unused after https://github.com/typescript-eslint/typescript-eslint/pull/5034

File: packages/website/package.json
Comment (auvred) [inline packages/website/package.json:32]: Unused after https://github.com/typescript-eslint/typescript-eslint/commit/acc1a43e0#diff-c84201550b7c349a26bad881654f20a00c4c0ce65ab023dd7b170efebaa03043

File: jest.config.js
Comment (JoshuaKGoldberg) [inline jest.config.js:3]: [Praise] Nice, +1 to getting `// @ts-check` in here.

File: knip.ts
Comment (JoshuaKGoldberg) [inline knip.ts:34]: ESLint actually does support `.mjs` now! We just never updated in our repo. `rm eslint.config.js` and these no longer need to be here.

```suggestion
```

File: knip.ts
Comment (JoshuaKGoldberg) [reply knip.ts:65]: Yeah this feels like a feature request for Knip: shouldn't it understand that `npx tsc` goes to TypeScript?

...separately, `-p` defaults to `tsconfig.json`. `yarn knip` and `yarn typecheck` pass happily for me when I remove the `-p tsconfig.json` altogether. Maybe a good reason to remove them?

File: knip.ts
Comment (JoshuaKGoldberg) [inline knip.ts:78]: Can you file/post something in Knip for Docusaurus support?

File: knip.ts
Comment (JoshuaKGoldberg) [inline knip.ts:90]: Maybe, so we don't have to keep a manual list? (pending Knip Docusaurus support)

```suggestion
        // imported in MDX docs
        'src/components/**/*.tsx',

        // used by Docusaurus
        'src/theme/**/index.tsx',
```

File: knip.ts
Comment (JoshuaKGoldberg) [inline knip.ts:131]: You know, as a followup I bet we could switch these over to plain `.ts` files, and remove the need for the `requireResolved` that's tripping up Knip... Not blocking at all IMO.

File: packages/utils/src/eslint-utils/index.ts
Comment (JoshuaKGoldberg) [reply packages/utils/src/eslint-utils/index.ts:2]: Probably yeah. The fact that nobody has reported anything makes me less motivated to - but, yeah, we should.

**Review by JoshuaKGoldberg — APPROVED** [review]
✂️ this is fantastic, nice job @auvred! Love the explanations throughout and all the removed stuff.

Nothing I commented is a blocker IMO. They can all be followup issues IMO.

File: knip.ts
Comment (auvred) [reply knip.ts:34]: It looks like `eslint.config.mjs` is not currently supported by `@nx/eslint`:

```
Linting "repo-tools"...

 NX   No ESLint configuration found in /<...>/typescript-eslint/packages/repo-tools.


Error: No ESLint configuration found in /<...>/typescript-eslint/packages/repo-tools.
    at CascadingConfigArrayFactory._finalizeConfigArray (/<...>/typescript-eslint/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:4000:19)
    at CascadingConfigArrayFactory.getConfigArrayForFile (/<...>/typescript-eslint/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3791:21)
    at FileEnumerator._iterateFilesRecursive (/<...>/typescript-eslint/node_modules/eslint/lib/cli-engine/file-enumerator.js:475:52)
    at _iterateFilesRecursive.next (<anonymous>)
    at FileEnumerator.iterateFiles (/<...>/typescript-eslint/node_modules/eslint/lib/cli-engine/file-enumerator.js:299:49)
    at iterateFiles.next (<anonymous>)
    at CLIEngine.executeOnFiles (/<...>/typescript-eslint/node_modules/eslint/lib/cli-engine/cli-engine.js:797:48)
    at ESLint.lintFiles (/<...>/typescript-eslint/node_modules/eslint/lib/eslint/eslint.js:551:23)
    at run (/<...>/typescript-eslint/node_modules/@nx/eslint/src/executors/lint/lint.impl.js:79:36)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

 NX   Running target lint for project repo-tools and 5 tasks it depends on failed

```

https://github.com/nrwl/nx/issues/22576

https://github.com/nrwl/nx/blob/f76b8c63c12ddfcb8de8d648cded026eec968ced/packages/eslint/src/executors/lint/lint.impl.ts#L46

File: knip.ts
Comment (auvred) [reply knip.ts:75]: :rocket: https://github.com/webpro/knip/issues/483#issuecomment-2073024263

File: packages/utils/src/eslint-utils/index.ts
Comment (auvred) [reply packages/utils/src/eslint-utils/index.ts:2]: Here it is:

https://github.com/typescript-eslint/typescript-eslint/issues/8975 -> https://github.com/typescript-eslint/typescript-eslint/pull/8976

File: knip.ts
Comment (JoshuaKGoldberg) [reply knip.ts:34]: Ah, right

**Review by JoshuaKGoldberg — APPROVED** [review]
🚀

**Comment by webpro** [comment]
Hey folks, this is great! I didn't check out the details of this PR, but feel free to drag me into this, love to help out a bit here. If there's feature requests or other annoying stuff in/from Knip I'd love to hear about it.

**Comment by auvred** [comment]
Integration tests failures are unrelated. See https://github.com/typescript-eslint/typescript-eslint/pull/9029#discussion_r1583877741:

> [8127873](https://github.com/typescript-eslint/typescript-eslint/commit/812787313626c927c506c58e00aff3857f2c163c) / #8640
> 
> This commit changed the `context.report` from `{ node, ... }` to `{ loc, ... }` Hence the `nodeType` is no longer visible in the output as the rule is not reporting against the node.


I think we can finally merge this PR!

**Comment by auvred** [comment]
Unfortunately, merging `main` two weeks ago automatically dismissed an existing approval, so I can't merge this PR (the PR authors can't approve their PR). Could someone please re-review and stamp this PR so we can merge it?

Every Monday, after the automatic release, there are a few merge conflicts due to bumped `@typescript-eslint/*` versions, and it's a bit tedious to resolve them every Monday :smile:

File: yarn.lock
Comment (auvred) [reply yarn.lock:1]: NP! Thank you for Renovate issue investigation!

Comment (JoshuaKGoldberg) [reply yarn.lock:1]: Yeah I think once someone takes the time to fix the test errors in https://github.com/typescript-eslint/typescript-eslint/pull/8589, we could probably keep it deduped forever. Which would be great.

**Review by JoshuaKGoldberg — APPROVED** [review]
💯 looks fantastic, thanks for pressing on this! Very excited to have the cross-file unused code detection.

Marking as `1 approval` to give you a chance to merge whenever you want. 😁

**Pull request merged**
